src/time/time.h

```diff
@@ -1,95 +1,114 @@
  */
-#define TIME_MAXDELTA   (time_t)(60 * 60 * 24 * 30)
+extern time_t time_start;
 
 /*
- * Time when process was started expressed as absolute unix timestamp
- * with a time_t type
- */
-time_t time_start;
-
-/*
- * We keep a cache of the current time of day in a global variable now
- * that is updated periodically by a timer event every second. This
- * saves us a bunch of time() system calls because we really only need
- * to get the time once a second, whereas there can be tens of thosands
- * of requests a second.
- *
- * Also keeping track of time as relative to server-start timestamp
- * instead of absolute unix timestamps gives us a space savings on
- * systems where sizeof(time_t) > sizeof(unsigned int)
- *
- * So, now actually holds 32-bit seconds since the server start time.
+ * Current time relative to process start. These are updated with each call
+ * to time_update().
  */
-rel_time_t now;
+extern rel_time_t now_sec;
+extern rel_time_fine_t now_ms;
+extern rel_time_fine_t now_us;
+extern rel_time_fine_t now_ns;
 
-/* Get the time the process started */
+/* Get the time the process was started */
 static inline time_t
 time_started(void)
 {
-    return time_start;
+    return __atomic_load_n(&time_start, __ATOMIC_RELAXED);
+}
+
+/* Get the current time in seconds since the process started */
+static inline rel_time_t
+time_now(void)
+{
+    return __atomic_load_n(&now_sec, __ATOMIC_RELAXED);
+}
+
+/* Get the current time in milliseconds since the process started */
+static inline rel_time_fine_t
+time_now_ms(void)
+{
+    return __atomic_load_n(&now_ms, __ATOMIC_RELAXED);
+}
+
+/* Get the current time in microseconds since the process started */
+static inline rel_time_fine_t
+time_now_us(void)
+{
+    return __atomic_load_n(&now_us, __ATOMIC_RELAXED);
 }
 
-/* Get the current absolute time (not time since process began) */
+/* Get the current time in nanoseconds since the process started */
+static inline rel_time_fine_t
+time_now_ns(void)
+{
+    return __atomic_load_n(&now_ns, __ATOMIC_RELAXED);
+}
+
+/* Get the current absolute time (not since process began) */
 static inline time_t
 time_now_abs(void)
 {
-    return time_start + (time_t)now;
+    return time_started() + time_now();
 }
 
-/* Get the current time (since process started) */
+/*
+ * time_reltime
+ *
+ * Depending on build options, time_reltime will take on different behaviors.
+ *
+ * In absolute timestamp only mode, t is treated as an absolute unix timestamp,
+ * and time_reltime will return the difference between the timestamp and the
+ * timestamp of when the server came up.
+ *
+ * In relative timestamp only mode, t is treated as a time relative to time now,
+ * and is converted into time since process start.
+ *
+ * In memcached compatibility mode, t is treated as it is in memcached, that is,
+ * if it is greater than 30 days, it is treated as an absolute timestamp;
+ * otherwise, it is treated as a relative timestamp.
+ */
+#if defined TIME_ABSOLUTE
```

Comment:
I'm generally in favor of command line option (over compile switch), because then we can use the same binary for different settings without recompiling.

Also, the naming here is backward-compatible but may be challenged given the expansion to include finer clocks.
Essentially, we want `time_reltime()` to give us a timestamp that has 0 set to when the process started, which is used to determine eviction/expiring. Here we can match the time_now signatures to indicate different granularity, e.g. `time_unix2process_sec(uint32_t t)`, `time_unix2process_ms(uint64_t t)`...

Reply:
Actually, the real problem with compile-time option is that if Redis wants one style of timestamp but twemcache wants another it would be really hard to compile them at once.